File: 002/scala/Solution.scala
import scala.math.sqrt
import scala.math.log
import scala.math.pow
import scala.math.floor

import projecteuler.Constants.PHI
import projecteuler.Constants.PSI
import projecteuler.Series.sumGeometric

object Solution {
  def main(args: Array[String]) {
    val p : Int = readInt()
    println(solve(p))
  }

  def solve(p: Int): Int = {
    val cte = 1/sqrt(5)
    val phi3 = PHI * PHI * PHI
    val psi3 = PSI * PSI * PSI
    val n = floor(log(p * sqrt(5))/log(PHI) + 1/2.0)
    val k = floor(n/3.0)
    val term1 = sumGeometric(phi3, k, phi3)
    val term2 = sumGeometric(psi3, k, psi3)
    (cte * (term1 + term2)).toInt
  }
}

File: 002/c++/solution.cc
#include <iostream> 
#include <cmath>
#include "factor.h"
#include "constant.h"
#include "series.h"

int solve(int p);

#ifndef TESTING
int main () {
  int p;
  std::cin >> p;
  std::cout << solve(p) << std::endl;
}
#endif

int solve (int p) {
  double cte = 1/sqrt(5);
  double h = phi();
  double s = psi();
  double phi3 = h * h * h;
  double psi3 = s * s * s;
  double n = floor(log(p * sqrt(5))/log(h) + 1/2.0);
  double k = floor(n/3.0);
  double term1 = sum_geometric(phi3, k, phi3);
  double term2 = sum_geometric(psi3, k, psi3);
  return (int) (cte * (term1 + term2));
}
